kyuubi-server/web-ui/src/components/monaco-editor/index.vue

```diff
@@ -0,0 +1,173 @@
+<!--
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+-->
+
+<template>
+  <div ref="codeEditBox" style="height: 100%" />
+</template>
+
+<script lang="ts" setup>
+  import * as monaco from 'monaco-editor'
+  import { format } from 'sql-formatter'
+  import jsonWorker from 'monaco-editor/esm/vs/language/json/json.worker?worker'
+  import cssWorker from 'monaco-editor/esm/vs/language/css/css.worker?worker'
+  import htmlWorker from 'monaco-editor/esm/vs/language/html/html.worker?worker'
+  import tsWorker from 'monaco-editor/esm/vs/language/typescript/ts.worker?worker'
+  import EditorWorker from 'monaco-editor/esm/vs/editor/editor.worker?worker'
+  import { editorProps } from './type'
+  import { useEditorStore } from '@/pinia/editor'
+  import { ref, toRaw, watch, onBeforeUnmount, onMounted } from 'vue'
+
+  // @ts-ignore: worker
+  self.MonacoEnvironment = {
+    getWorker(_: string, label: string) {
+      if (label === 'json') {
+        return new jsonWorker()
+      }
+      if (['css', 'scss', 'less'].includes(label)) {
+        return new cssWorker()
+      }
+      if (['html', 'handlebars', 'razor'].includes(label)) {
+        return new htmlWorker()
+      }
+      if (['typescript', 'javascript'].includes(label)) {
+        return new tsWorker()
+      }
+      return new EditorWorker()
+    }
+  }
+
+  const props = defineProps(editorProps)
+  const emit = defineEmits([
+    'update:modelValue',
+    'change',
+    'editorMounted',
+    'editorSave'
+  ])
+
+  const editorStore = useEditorStore()
+  const monacoEditorThemeRef = ref(
+    editorStore.getCurrentTheme === 'dark' ? 'vs-dark' : 'vs'
+  )
+  let editor: monaco.editor.IStandaloneCodeEditor
+  const codeEditBox = ref()
+  const init = () => {
+    monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions({
+      noSemanticValidation: true,
+      noSyntaxValidation: false
+    })
+    monaco.languages.typescript.javascriptDefaults.setCompilerOptions({
+      target: monaco.languages.typescript.ScriptTarget.ES2020,
+      allowNonTsExtensions: true
+    })
+    monaco.languages.registerCompletionItemProvider('sql', {
```

Comment:
must we write such a completion hepler by hand?

Reply:
Never mind, this is just a simple version of sql functionality

Reply:
I thought there may be a popular third-party implementation, seems not ...